File: backend/app/services/bank_field_schemas.py
"""ZR-PAY-LINK-003 Wireframe D: "Account details [ secure jurisdiction-
specific fields ]" -- per-country structured bank-detail field sets (UK
sort code + account number, EU IBAN, US routing + account number), same
shape and posture as app/services/bank_identifiers.py's own per-jurisdiction
format registry: public banking-standard technical fact (like a postal
code's shape), not legal/regulatory content, so adding a new country here
doesn't need counsel review the way a MarketPolicyPack would.

Deliberately fails OPEN, not closed, for an unrecognized country --
resolve_bank_field_schema always returns a usable schema (the generic
fallback) rather than raising. This is a UX/data-quality aid, not a legal
gate like the Listing Fee/MarketPolicyPack checks; it must never brick a
host's ability to submit payment instructions just because nobody's added
their country's field set yet."""

from __future__ import annotations

import re
from dataclasses import dataclass


@dataclass(frozen=True)
class BankField:
    key: str
    label: str
    pattern: re.Pattern[str]
    hint: str = ""


@dataclass(frozen=True)
class BankFieldSchema:
    fields: tuple[BankField, ...]
    # Which field's value backs account_identifier_last4/masked display --
    # always the one a tenant would recognize as "the account number".
    primary_field_key: str


# UK sort code: 6 digits, conventionally written in 2-2-2 groups either bare
# or dash-separated (e.g. "123456" or "12-34-56") -- same regex as
# bank_identifiers.py's own UK sort-code format.
_UK_SORT_CODE = re.compile(r"^\d{2}-?\d{2}-?\d{2}$")
_UK_ACCOUNT_NUMBER = re.compile(r"^\d{8}$")
_US_ROUTING_NUMBER = re.compile(r"^\d{9}$")
_US_ACCOUNT_NUMBER = re.compile(r"^\d{4,17}$")
# IBAN: 2-letter country code + 2 check digits + up to 30 alphanumeric BBAN
# characters -- the general ISO 13616 shape, not a per-country checksum
# validator (that's a heavier lift this build doesn't need for a UX-level
# field-shape check).
_IBAN = re.compile(r"^[A-Z]{2}\d{2}[A-Z0-9]{1,30}$")
_GENERIC_IDENTIFIER = re.compile(r"^.{4,64}$")

_GB_SCHEMA = BankFieldSchema(
    fields=(
        BankField("sort_code", "Sort code", _UK_SORT_CODE, "6 digits, e.g. 12-34-56"),
        BankField("account_number", "Account number", _UK_ACCOUNT_NUMBER, "8 digits"),
    ),
    primary_field_key="account_number",
)
_US_SCHEMA = BankFieldSchema(
    fields=(
        BankField("routing_number", "Routing number", _US_ROUTING_NUMBER, "9 digits"),
        BankField("account_number", "Account number", _US_ACCOUNT_NUMBER, "4-17 digits"),
    ),
    primary_field_key="account_number",
)
_IBAN_SCHEMA = BankFieldSchema(
    fields=(BankField("iban", "IBAN", _IBAN, "e.g. DE89370400440532013000"),),
    primary_field_key="iban",
)
# The pre-existing single-generic-field behavior, preserved as the fallback
# for any country not explicitly listed below.
_FALLBACK_SCHEMA = BankFieldSchema(
    fields=(BankField("account_identifier", "Account / payment ID", _GENERIC_IDENTIFIER, "At least 4 characters"),),
    primary_field_key="account_identifier",
)

# A small, explicitly extensible starting set -- same "add a country, don't
# invent a format" posture as bank_identifiers.py. Not exhaustive of every
# IBAN-using country; unlisted ones fall through to the generic schema
# below rather than being silently blocked.
_IBAN_COUNTRIES = frozenset({"DE", "FR", "ES", "IT", "NL", "IE", "PT", "BE"})

BANK_FIELD_SCHEMAS: dict[str, BankFieldSchema] = {
    "GB": _GB_SCHEMA,
    "US": _US_SCHEMA,
    **{code: _IBAN_SCHEMA for code in _IBAN_COUNTRIES},
}


def resolve_bank_field_schema(country_code: str, method: str = "BANK_TRANSFER") -> BankFieldSchema:
    """Country-specific structured fields (sort code, IBAN, routing number)
    only ever make sense for BANK_TRANSFER -- a CASH/CARD/OTHER instruction
    has no bank routing details at all, forcing a UK sort code onto a cash
    payment would be nonsensical, not just unhelpful. Every non-bank-
    transfer method always gets the generic single-field fallback,
    regardless of country_code."""
    if method != "BANK_TRANSFER":
        return _FALLBACK_SCHEMA
    return BANK_FIELD_SCHEMAS.get(country_code.strip().upper(), _FALLBACK_SCHEMA)


def validate_bank_details(country_code: str, bank_details: dict[str, str], method: str = "BANK_TRANSFER") -> None:
    """Raises ValueError (caller converts to HTTPException 400) on a missing
    or malformed field. Extra keys not in the schema are ignored rather
    than rejected -- lets the frontend submit a superset without this
    validator needing to change in lockstep."""
    schema = resolve_bank_field_schema(country_code, method)
    for field in schema.fields:
        value = (bank_details.get(field.key) or "").strip()
        if not value:
            raise ValueError(f"{field.label} is required")
        if not field.pattern.match(value):
            raise ValueError(f"{field.label} is not in a valid format ({field.hint})" if field.hint else f"{field.label} is not in a valid format")

File: src/lib/bankFieldSchemas.ts
/** ZR-PAY-LINK-003 Wireframe D: "Country / region" + "Account details
 *  [ secure jurisdiction-specific fields ]" -- mirrors
 *  backend/app/services/bank_field_schemas.py's registry exactly (same
 *  country list, same field keys/labels/patterns) so client-side
 *  validation matches what the server will actually accept. The server is
 *  still the authority -- this is only for immediate form feedback. */

export interface BankFieldDef {
  key: string;
  label: string;
  pattern: RegExp;
  hint: string;
}

export interface BankFieldSchema {
  fields: BankFieldDef[];
  primaryFieldKey: string;
}

const UK_SORT_CODE = /^\d{2}-?\d{2}-?\d{2}$/;
const UK_ACCOUNT_NUMBER = /^\d{8}$/;
const US_ROUTING_NUMBER = /^\d{9}$/;
const US_ACCOUNT_NUMBER = /^\d{4,17}$/;
const IBAN = /^[A-Z]{2}\d{2}[A-Z0-9]{1,30}$/;
const GENERIC_IDENTIFIER = /^.{4,64}$/;

const GB_SCHEMA: BankFieldSchema = {
  fields: [
    { key: "sort_code", label: "Sort code", pattern: UK_SORT_CODE, hint: "6 digits, e.g. 12-34-56" },
    { key: "account_number", label: "Account number", pattern: UK_ACCOUNT_NUMBER, hint: "8 digits" },
  ],
  primaryFieldKey: "account_number",
};
const US_SCHEMA: BankFieldSchema = {
  fields: [
    { key: "routing_number", label: "Routing number", pattern: US_ROUTING_NUMBER, hint: "9 digits" },
    { key: "account_number", label: "Account number", pattern: US_ACCOUNT_NUMBER, hint: "4-17 digits" },
  ],
  primaryFieldKey: "account_number",
};
const IBAN_SCHEMA: BankFieldSchema = {
  fields: [{ key: "iban", label: "IBAN", pattern: IBAN, hint: "e.g. DE89370400440532013000" }],
  primaryFieldKey: "iban",
};
export const FALLBACK_SCHEMA: BankFieldSchema = {
  fields: [{ key: "account_identifier", label: "Account / payment ID", pattern: GENERIC_IDENTIFIER, hint: "At least 4 characters" }],
  primaryFieldKey: "account_identifier",
};

const IBAN_COUNTRIES = ["DE", "FR", "ES", "IT", "NL", "IE", "PT", "BE"] as const;

export const BANK_FIELD_SCHEMAS: Record<string, BankFieldSchema> = {
  GB: GB_SCHEMA,
  US: US_SCHEMA,
  ...Object.fromEntries(IBAN_COUNTRIES.map((code) => [code, IBAN_SCHEMA])),
};

/** Options for the "Country" select -- GB/US first (most common for this
 *  build's own markets), then the IBAN countries, then a generic "Other"
 *  that resolves to FALLBACK_SCHEMA (same fail-open, not fail-closed,
 *  posture as the backend registry -- an unlisted country must never
 *  block submission). */
export const COUNTRY_OPTIONS: { code: string; label: string }[] = [
  { code: "GB", label: "United Kingdom" },
  { code: "US", label: "United States" },
  { code: "DE", label: "Germany" },
  { code: "FR", label: "France" },
  { code: "ES", label: "Spain" },
  { code: "IT", label: "Italy" },
  { code: "NL", label: "Netherlands" },
  { code: "IE", label: "Ireland" },
  { code: "PT", label: "Portugal" },
  { code: "BE", label: "Belgium" },
  { code: "OTHER", label: "Other" },
];

/** Country-specific structured fields only ever make sense for
 *  BANK_TRANSFER -- a CASH/CARD/OTHER instruction has no bank routing
 *  details, so every other method always gets the generic single-field
 *  fallback regardless of country. Mirrors
 *  backend/app/services/bank_field_schemas.py:resolve_bank_field_schema's
 *  own method check exactly. */
export function resolveBankFieldSchema(countryCode: string, method: string = "BANK_TRANSFER"): BankFieldSchema {
  if (method !== "BANK_TRANSFER") return FALLBACK_SCHEMA;
  return BANK_FIELD_SCHEMAS[countryCode.toUpperCase()] ?? FALLBACK_SCHEMA;
}
